internal/npm_install/generate_build_file.js

```diff
@@ -0,0 +1,276 @@
+'use strict';
+
+const fs = require('fs');
+const path = require('path');
+
+const DEBUG = console.error;
+
+const generatedHeader = `"""Generated file from yarn_install rule.
+See $(bazel info output_base)/external/build_bazel_rules_nodejs/internal/npm_install/generate_build_file.js
+"""
+
+# All rules in other repositories can use these targets
+package(default_visibility = ["//visibility:public"])
+
+load("@build_bazel_rules_nodejs//:defs.bzl", "nodejs_binary")
+
+`;
+
+const allFilegroup = `
+# Export the entire node_modules directory in one catch-all rule.
+# NB. this has bad performance implications if there are many files.
+filegroup(
+    name = "node_modules",
+    srcs = glob(
+        include = ["node_modules/**/*"],
+        # Files with spaces in the name are not legal Bazel labels
+        exclude = [
+          "node_modules/**/* */**",
+          "node_modules/**/* *",
+        ],
+    ),
+)
+
+# Export a trimmed version of the node_modules filegroup that includes
+# only js, d.ts and json files as well as the .bin folder. This can
+# be used in some cases to improve performance by reducing the number
+# of runfiles. The recommended approach to reducing performance
+# is to use fine grained deps such as ["@npm//:a", "@npm://b", ...]
+filegroup(
+    name = "node_modules_trimmed",
+    srcs = glob(
+        include = [
+          "node_modules/**/*.js",
+          "node_modules/**/*.d.ts",
+          "node_modules/**/*.json",
+        ],
+        # Files with spaces in the name are not legal Bazel labels
+        exclude = [
+          "node_modules/**/* */**",
+          "node_modules/**/* *",
+        ],
+    ) + glob(["node_modules/.bin/*"]),
+)
+
+# Export a node_modules_none file group which is useful
+# for nodejs_binary targets with no npm dependencies
+# but that still need to specify a node_modules attribute
+filegroup(
+    name = "node_modules_none",
+    srcs = [],
+)
+`
+
+
+if (require.main === module) {
+  main();
+}
+
+/**
+ * Main entrypoint.
+ * Write BUILD file.
+ */
+function main() {
+  // Scan the node_modules directory and find all top-level ('foo') or scoped (@bar/baz)
+  // packages, i.e. folders which contain a package.json file then parse the package.json
+  // files to collect the metadata
+  let pkgDirs = [
+    ...getPackagesIn('node_modules'),
+    ...findScopes('node_modules')
+        .map(scope => getPackagesIn(scope))
+        .reduce((a, b) => a.concat(b), [])
+  ];
+  let nestedPkgDirs = [];
+  pkgDirs.forEach(d => {
```

Comment:
looking for nested packages is necessary but the way you've done it here isn't enough.

node modules can be arbitrarily nested, for example you could have:
```
node_modules/
    bar/
     baz/
    foo/
        node_modules/
            bar/
                 node_modules/
                     baz/
```

It is important to traverse all the way down because the `baz` that resides three levels deep may have different dependencies than the `baz` that resides at the top-level node_modules (due to version differences). In this case, the transitive dependencies listed in the generated build file will not be accurate, since the `baz` found in pkgDirs will be the top-level one and not contain the correct dependencies.

I think you can just continue recursing down the found packages until none of them add any packages. Eventually the leaf packages will not have a nested node_modules directory.

Reply:
@kitofans Thats a good point. I'll update the logic to handle arbitrary nested node_modules depth. Thanks!